<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta name="description" content="None">
  
  
  <link rel="shortcut icon" href="img/favicon.ico">
  <title>Learning the BAM format</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="css/theme.css" />
  <link rel="stylesheet" href="css/theme_extra.css" />
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/styles/github.min.css" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "Home";
    var mkdocs_page_input_path = "index.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="js/jquery-2.1.1.min.js" defer></script>
  <script src="js/modernizr-2.8.3.min.js" defer></script>
  <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/10.5.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href="." class="icon icon-home"> Learning the BAM format</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="./search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href=".">Home</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#introduction">Introduction</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#installing-samtools">Installing SAMtools</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#basic-usage">Basic usage</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#viewing">Viewing</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#converting-a-sam-file-to-a-bam-file">Converting a SAM file to a BAM file</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#converting-a-bam-file-to-a-cram-file">Converting a BAM file to a CRAM file</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#sorting-a-sambam-file">Sorting a SAM/BAM file</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#creating-a-bam-index-file">Creating a BAM index file</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#adding-read-groups">Adding read groups</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#interpreting-the-bam-flags">Interpreting the BAM flags</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#filtering-unmapped-reads">Filtering unmapped reads</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#proper-pair">Proper pair</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#extracting-entries-mapping-to-a-specific-loci">Extracting entries mapping to a specific loci</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#extracting-only-the-first-read-from-paired-end-bam-files">Extracting only the first read from paired end BAM files</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#stats">Stats</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#samtools-calmdfillmd">samtools calmd/fillmd</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#creating-fastq-files-from-a-bam-file">Creating FASTQ files from a BAM file</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#random-subsampling-of-bam-file">Random subsampling of BAM file</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#count-number-of-reads">Count number of reads</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#obtaining-genomic-sequence">Obtaining genomic sequence</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#comparing-bam-files">Comparing BAM files</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#converting-reference-names">Converting reference names</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#coverage">Coverage</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#stargazers-over-time">Stargazers over time</a>
    </li>
    </ul>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href=".">Learning the BAM format</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".">Docs</a> &raquo;</li>
    
      
    
    <li>Home</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  
  <hr/>
</div>

          <div role="main">
            <div class="section">
              
                <h1 id="learning-the-bam-format">Learning the BAM format</h1>
<h2 id="introduction">Introduction</h2>
<p><img alt="Build
README" src="https://github.com/davetang/learning_bam_file/actions/workflows/create_readme.yml/badge.svg" /></p>
<p>SAMtools provides various (sub)tools for manipulating alignments in the
SAM/BAM format. The SAM (Sequence Alignment/Map) format (BAM is just the
binary form of SAM) is currently the <em>de facto</em> standard for storing
large nucleotide sequence alignments. If you are working with
high-throughput sequencing data, at some point you will probably have to
deal with SAM/BAM files, so familiarise yourself with them! For the
latest information on SAMtools, please refer to the <a href="https://github.com/samtools/samtools/releases">release
notes</a>.</p>
<p>The examples in this README use the <code>ERR188273_chrX.bam</code> BAM file
(stored in the <code>eg</code> folder) generated as per
<a href="https://github.com/davetang/rnaseq">https://github.com/davetang/rnaseq</a> using the HISAT2 + StringTie2
RNA-seq pipeline. This README is generated using the <code>create_readme.sh</code>
script; if you want to generate this file yourself, please use <a href="https://hub.docker.com/repository/docker/davetang/r_build">this
Docker image</a>
and the <code>Makefile</code> in this directory. For example:</p>
<pre><code class="language-bash"># clone this repo
git clone https://github.com/davetang/learning_bam_file.git
cd learning_bam_file

docker pull davetang/r_build:4.1.2
docker run --rm -it -v $(pwd):/work davetang/r_build:4.1.2 /bin/bash

# inside the Docker container
make
</code></pre>
<h2 id="installing-samtools">Installing SAMtools</h2>
<p>For installing SAMtools, I recommend using <code>Conda</code> and the <a href="https://anaconda.org/bioconda/samtools">Bioconda
samtools package</a>. I also
recommend using
<a href="https://docs.conda.io/en/latest/miniconda.html">Miniconda</a> instead of
Anaconda because Anaconda comes with a lot of tools/packages that you
will probably not use. I wrote a <a href="https://davetang.github.io/reproducible_bioinformatics/conda.html">short introduction to
Conda</a>
if you want to find learn more.</p>
<p>Once you have installed Miniconda, you can install SAMtools as follows:</p>
<pre><code class="language-bash">conda install -c bioconda samtools
</code></pre>
<p>Otherwise you can download the source and compile it yourself; change
<code>dir</code> to the location you want <code>samtools</code> to be installed. <code>samtools</code>
will be installed in <code>${dir}/bin</code>, so make sure this is in your <code>$PATH</code>.</p>
<pre><code class="language-bash">#!/usr/bin/env bash

set -euo pipefail

ver=1.15
tool=samtools
url=https://github.com/samtools/${tool}/releases/download/${ver}/${tool}-${ver}.tar.bz2
dir=${HOME}/local

wget ${url}
tar xjf ${tool}-${ver}.tar.bz2
cd ${tool}-${ver}
./configure --prefix=${dir}
make &amp;&amp; make install
cd ..

rm -rf ${tool}-${ver} ${tool}-${ver}.tar.bz2

&gt;&amp;2 echo Done
exit 0
</code></pre>
<h2 id="basic-usage">Basic usage</h2>
<p>If you run <code>samtools</code> on the terminal without any parameters or with
<code>--help</code>, all the available utilities are listed:</p>
<pre><code class="language-bash">samtools --help
</code></pre>
<pre><code>## 
## Program: samtools (Tools for alignments in the SAM format)
## Version: 1.15 (using htslib 1.15)
## 
## Usage:   samtools &lt;command&gt; [options]
## 
## Commands:
##   -- Indexing
##      dict           create a sequence dictionary file
##      faidx          index/extract FASTA
##      fqidx          index/extract FASTQ
##      index          index alignment
## 
##   -- Editing
##      calmd          recalculate MD/NM tags and '=' bases
##      fixmate        fix mate information
##      reheader       replace BAM header
##      targetcut      cut fosmid regions (for fosmid pool only)
##      addreplacerg   adds or replaces RG tags
##      markdup        mark duplicates
##      ampliconclip   clip oligos from the end of reads
## 
##   -- File operations
##      collate        shuffle and group alignments by name
##      cat            concatenate BAMs
##      consensus      produce a consensus Pileup/FASTA/FASTQ
##      merge          merge sorted alignments
##      mpileup        multi-way pileup
##      sort           sort alignment file
##      split          splits a file by read group
##      quickcheck     quickly check if SAM/BAM/CRAM file appears intact
##      fastq          converts a BAM to a FASTQ
##      fasta          converts a BAM to a FASTA
##      import         Converts FASTA or FASTQ files to SAM/BAM/CRAM
## 
##   -- Statistics
##      bedcov         read depth per BED region
##      coverage       alignment depth and percent coverage
##      depth          compute the depth
##      flagstat       simple stats
##      idxstats       BAM index stats
##      phase          phase heterozygotes
##      stats          generate stats (former bamcheck)
##      ampliconstats  generate amplicon specific stats
## 
##   -- Viewing
##      flags          explain BAM flags
##      head           header viewer
##      tview          text alignment viewer
##      view           SAM&lt;-&gt;BAM&lt;-&gt;CRAM conversion
##      depad          convert padded BAM to unpadded BAM
##      samples        list the samples in a set of SAM/BAM/CRAM files
## 
##   -- Misc
##      help [cmd]     display this help message or help for [cmd]
##      version        detailed version information
</code></pre>
<h2 id="viewing">Viewing</h2>
<p>Use <a href="https://github.com/bioSyntax/bioSyntax">bioSyntax</a> to prettify your
output.</p>
<pre><code class="language-bash">samtools view aln.bam | sam-less
</code></pre>
<p><img alt="bioSyntax" src="img/sam_less.png" /></p>
<h2 id="converting-a-sam-file-to-a-bam-file">Converting a SAM file to a BAM file</h2>
<p>A BAM file is just a SAM file but stored in binary format; you should
always convert your SAM files into BAM format since they are smaller in
size and are faster to manipulate.</p>
<p>I don’t have a SAM file in the example folder, so let’s create one and
check out the first ten lines. Note: remember to use <code>-h</code> to ensure the
SAM file contains the sequence header information. Generally, I
recommend storing only sorted BAM files as they use even less disk space
and are faster to process.</p>
<pre><code class="language-bash">samtools view -h eg/ERR188273_chrX.bam &gt; eg/ERR188273_chrX.sam
</code></pre>
<p>Notice that the SAM file is much larger than the BAM file.</p>
<p>Size of SAM file.</p>
<pre><code class="language-bash">ls -lh eg/ERR188273_chrX.sam
</code></pre>
<pre><code>## -rw-r--r-- 1 root root 321M Apr 18 08:35 eg/ERR188273_chrX.sam
</code></pre>
<p>Size of BAM file.</p>
<pre><code class="language-bash">ls -lh eg/ERR188273_chrX.bam
</code></pre>
<pre><code>## -rw-r--r-- 1 root root 67M Apr 18 08:33 eg/ERR188273_chrX.bam
</code></pre>
<p>We can use <code>head</code> to view a SAM file.</p>
<pre><code class="language-bash">head eg/ERR188273_chrX.sam
</code></pre>
<pre><code>## @HD  VN:1.0  SO:coordinate
## @SQ  SN:chrX LN:156040895
## @PG  ID:hisat2   PN:hisat2   VN:2.2.0    CL:"/Users/dtang/github/rnaseq/hisat2/../src/hisat2-2.2.0/hisat2-align-s --wrapper basic-0 --dta -p 4 -x ../raw/chrX_data/indexes/chrX_tran -1 /tmp/4195.inpipe1 -2 /tmp/4195.inpipe2"
## @PG  ID:samtools PN:samtools PP:hisat2   VN:1.15 CL:samtools view -h eg/ERR188273_chrX.bam
## ERR188273.4711308    73  chrX    21649   0   5S70M   =   21649   0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.4711308    133 chrX    21649   0   *   =   21649   0   CTACAGGTGCCCGCCACCATGCCCAGCTAATTTTTTTTGTATTTTTAGTAGAGATGGGGTTTCACTGTGTTGGCC CB@FDFFFHHGFHIJJJJIIIIIIIGGGIJGIIJJJJJJFFHIIIIGECHEHHGGHHFF?AACCDDDDDDDDBCD YT:Z:UP
## ERR188273.4711308    329 chrX    233717  0   5S70M   =   233717  0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.14904746   99  chrX    251271  60  75M =   251317  121 GAAAAATGGGCCCAGGGGACCGGCGCTCAGCATACAGAGGACCCGCGCCGGCACCTGCCTCTGAGTTCCCTTAGT @@&lt;DDDDDFB&gt;HHEGIIGAGIIIBGIIG@FECH&lt;F@GIIFAE=?BCBBCBBB5@&lt;?CBBCCCCAACDCCCCCCCC AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YS:i:-2 YT:Z:CP NH:i:1
## ERR188273.14904746   147 chrX    251317  60  75M =   251271  -121    GCCGGCCCCTGCCTCTGAGTTCCCTTAGTACTTATTGATCATTATCGGGGAGAGGGGGATGTGGCAGGACAATAG #######B?DAHC@EGIIGGEHHGC@GFBFCEGFCIGG@EG@@H&lt;JIEHEF@IGEHGHIIHFGHDDFDDDDD?&lt;B AS:i:-2 ZS:i:-7 XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:6A68   YS:i:0  YT:Z:CP NH:i:1
## ERR188273.5849805    163 chrX    265951  1   75M =   266022  146 CGGGTTCACGCCATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGACTACAGGCGCCCGCCACCACGCCCGGCTAA @CCFDFFFHHHHHJJJJJJFJJJJJIJIJJJJJJJJGHIJJJJJEHIJIJGIIJJJHHFFDDEDDDDDDDDDDDD AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YS:i:0  YT:Z:CP NH:i:2
</code></pre>
<p>The lines starting with an “@” symbol contains the header information.
The @SQ tag is the reference sequence dictionary; SN refers to the
reference sequence name and LN refers to the reference sequence length.
If you don’t see lines starting with the “@” symbol, the header
information is probably missing. You can generate this information again
by running the command below, where <code>ref.fa</code> is the reference FASTA file
used to map the reads.</p>
<pre><code class="language-bash">samtools view -bT sequence/ref.fa aln.sam &gt; aln.bam
</code></pre>
<p>If the header information is available, we can convert a SAM file into
BAM by using <code>samtools view -b</code>. In newer versions of SAMtools, the
input format is auto-detected, so we no longer need the <code>-S</code> parameter.</p>
<pre><code class="language-bash">samtools view -b eg/ERR188273_chrX.sam &gt; eg/my.bam
</code></pre>
<h2 id="converting-a-bam-file-to-a-cram-file">Converting a BAM file to a CRAM file</h2>
<p>The CRAM format is even more compact. Use <code>samtools view</code> with the <code>-T</code>
and <code>-C</code> arguments to convert a BAM file into CRAM.</p>
<pre><code class="language-bash">samtools view -T genome/chrX.fa -C -o eg/ERR188273_chrX.cram eg/ERR188273_chrX.bam

ls -lh eg/ERR188273_chrX.[sbcr]*am
</code></pre>
<pre><code>## -rw-r--r-- 1 root root  67M Apr 18 08:33 eg/ERR188273_chrX.bam
## -rw-r--r-- 1 root root  40M Apr 18 08:35 eg/ERR188273_chrX.cram
## -rw-r--r-- 1 root root 321M Apr 18 08:35 eg/ERR188273_chrX.sam
</code></pre>
<p>You can use <code>samtools view</code> to view a CRAM file just as you would for a
BAM file.</p>
<pre><code class="language-bash">samtools view eg/ERR188273_chrX.cram | head
</code></pre>
<pre><code>## ERR188273.4711308    73  chrX    21649   0   5S70M   =   21649   0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  YT:Z:UP NH:i:2  MD:Z:70 NM:i:0
## ERR188273.4711308    133 chrX    21649   0   *   =   21649   0   CTACAGGTGCCCGCCACCATGCCCAGCTAATTTTTTTTGTATTTTTAGTAGAGATGGGGTTTCACTGTGTTGGCC CB@FDFFFHHGFHIJJJJIIIIIIIGGGIJGIIJJJJJJFFHIIIIGECHEHHGGHHFF?AACCDDDDDDDDBCD YT:Z:UP
## ERR188273.4711308    329 chrX    233717  0   5S70M   =   233717  0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  YT:Z:UP NH:i:2  MD:Z:70 NM:i:0
## ERR188273.14904746   99  chrX    251271  60  75M =   251317  121 GAAAAATGGGCCCAGGGGACCGGCGCTCAGCATACAGAGGACCCGCGCCGGCACCTGCCTCTGAGTTCCCTTAGT @@&lt;DDDDDFB&gt;HHEGIIGAGIIIBGIIG@FECH&lt;F@GIIFAE=?BCBBCBBB5@&lt;?CBBCCCCAACDCCCCCCCC AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  YS:i:-2 YT:Z:CP NH:i:1  MD:Z:75 NM:i:0
## ERR188273.14904746   147 chrX    251317  60  75M =   251271  -121    GCCGGCCCCTGCCTCTGAGTTCCCTTAGTACTTATTGATCATTATCGGGGAGAGGGGGATGTGGCAGGACAATAG #######B?DAHC@EGIIGGEHHGC@GFBFCEGFCIGG@EG@@H&lt;JIEHEF@IGEHGHIIHFGHDDFDDDDD?&lt;B AS:i:-2 ZS:i:-7 XN:i:0  XM:i:1  XO:i:0  XG:i:0  YS:i:0  YT:Z:CP NH:i:1  MD:Z:6A68   NM:i:1
## ERR188273.5849805    163 chrX    265951  1   75M =   266022  146 CGGGTTCACGCCATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGACTACAGGCGCCCGCCACCACGCCCGGCTAA @CCFDFFFHHHHHJJJJJJFJJJJJIJIJJJJJJJJGHIJJJJJEHIJIJGIIJJJHHFFDDEDDDDDDDDDDDD AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  YS:i:0  YT:Z:CP NH:i:2  MD:Z:75 NM:i:0
## ERR188273.1232356    369 chrX    265984  1   75M =   118343251   0   GAGTAGCTGGGACTACAGGCGCCCGCCACCACGCCCGGCTAATTTTTTGTATTTTTAGTAGAGACGGGGTTTCAC @@CA@B&gt;DC&gt;&gt;+@::8-755-BBBFDDEHHBGGEGHEEIJIIGIJJIGEIIIJJJIIJJIGGHHHGGFFFFF@@C AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  YT:Z:UP NH:i:10 MD:Z:75 NM:i:0
## ERR188273.5927795    385 chrX    265991  1   75M =   114048277   0   TGGGACTACAGGCGCCCGCCACCACGCCCGGCTAATTTTTTGTATTTTTAGTAGAGACGGGGTTTCACCGTGTTA =?BB??BD?FBHHBEAE@CDGG@HH=FA@GEGE;FGACCHBE6?A=ACE9)7@DCE&gt;&gt;5'3=338:;:&gt;2&lt;AA?: AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  YT:Z:UP NH:i:10 MD:Z:75 NM:i:0
## ERR188273.5849805    83  chrX    266022  1   75M =   265951  -146    CTAATTTTTTGTATTTTTAGTAGAGACGGGGTTTCACCGTGTTAGCCAGGATGGTGTCGATCTCCTGACCTCGTG DDDDDDDEEEEEDBFEDGHHHHHHJHIJJIGIGHFBJJIHGJJIIJJJJJJJJIGIJJJJJJHHHHHDFFFFCCB AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  YS:i:0  YT:Z:CP NH:i:2  MD:Z:75 NM:i:0
## ERR188273.13655123   113 chrX    266022  1   75M =   118343234   0   CTAATTTTTTGTATTTTTAGTAGAGACGGGGTTTCACCGTGTTAGCCAGGATGGTGTCGATCTCCTGACCTCGTG AACBBBACCCC&gt;;3?BCFFEEHHHEEGIGGHAGFBBHFBHHEHCG@&lt;@ABG??@@?BB9GBGAFFD&lt;&lt;DDAD@@@ AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  YT:Z:UP NH:i:2  MD:Z:75 NM:i:0
</code></pre>
<p>I have an <a href="https://davetang.org/muse/2014/09/26/bam-to-cram/">old blog
post</a> on the CRAM
format.</p>
<h2 id="sorting-a-sambam-file">Sorting a SAM/BAM file</h2>
<p>Many downstream tools require sorted BAM files and since they are
slightly more compact than unsorted BAM files, you should always sorted
BAM files. In SAMtools version 1.3 or newer, you can directly generate a
sorted BAM file from a SAM file.</p>
<pre><code class="language-bash">samtools sort eg/ERR188273_chrX.sam -o eg/sorted.bam
ls -l eg/ERR188273_chrX.bam
ls -l eg/sorted.bam
</code></pre>
<pre><code>## -rw-r--r-- 1 root root 69983526 Apr 18 08:33 eg/ERR188273_chrX.bam
## -rw-r--r-- 1 root root 69983598 Apr 18 08:35 eg/sorted.bam
</code></pre>
<p>You should use use additional threads (if they are available) to speed
up sorting; to use four threads, use <code>-@ 4</code>.</p>
<p>Time taken using one thread (default).</p>
<pre><code class="language-bash">time samtools sort eg/ERR188273_chrX.sam -o eg/sorted.bam
</code></pre>
<pre><code>## 
## real 0m10.378s
## user 0m10.031s
## sys  0m0.276s
</code></pre>
<p>Time taken using four threads.</p>
<pre><code class="language-bash">time samtools sort -@ 4 eg/ERR188273_chrX.sam -o eg/sorted.bam
</code></pre>
<pre><code>## [bam_sort_core] merging from 0 files and 4 in-memory blocks...
## 
## real 0m5.493s
## user 0m10.205s
## sys  0m0.300s
</code></pre>
<p>Many of the SAMtools subtools can use additional threads, so make use of
them if you have the resources!</p>
<h2 id="creating-a-bam-index-file">Creating a BAM index file</h2>
<p>Various tools require BAM index files, such as IGV, which is a tool that
can be used for visualising BAM files.</p>
<pre><code class="language-bash">samtools index eg/ERR188273_chrX.bam
</code></pre>
<h2 id="adding-read-groups">Adding read groups</h2>
<p>Some tools like GATK and Picard require <a href="https://gatk.broadinstitute.org/hc/en-us/articles/360035890671-Read-groups">read
groups</a>
(RG). You can add or replace read groups using <code>samtools addreplacerg</code>.</p>
<pre><code class="language-bash">samtools addreplacerg -r &quot;@RG\tID:ERR188273\tSM:ERR188273\tPL:illumina&quot; -o eg/ERR188273_chrX_rg.bam eg/ERR188273_chrX.bam
samtools head eg/ERR188273_chrX_rg.bam
</code></pre>
<pre><code>## @HD  VN:1.0  SO:coordinate
## @SQ  SN:chrX LN:156040895
## @PG  ID:hisat2   PN:hisat2   VN:2.2.0    CL:"/Users/dtang/github/rnaseq/hisat2/../src/hisat2-2.2.0/hisat2-align-s --wrapper basic-0 --dta -p 4 -x ../raw/chrX_data/indexes/chrX_tran -1 /tmp/4195.inpipe1 -2 /tmp/4195.inpipe2"
## @PG  ID:samtools PN:samtools PP:hisat2   VN:1.15 CL:samtools addreplacerg -r @RG\tID:ERR188273\tSM:ERR188273\tPL:illumina -o eg/ERR188273_chrX_rg.bam eg/ERR188273_chrX.bam
## @RG  ID:ERR188273    SM:ERR188273    PL:illumina
</code></pre>
<p>If you want to replace existing read groups, just use the same command.</p>
<pre><code class="language-bash">samtools addreplacerg -r &quot;@RG\tID:ERR188273_2\tSM:ERR188273_2\tPL:illumina_2&quot; -o eg/ERR188273_chrX_rg2.bam eg/ERR188273_chrX_rg.bam
samtools head eg/ERR188273_chrX_rg2.bam
</code></pre>
<pre><code>## @HD  VN:1.0  SO:coordinate
## @SQ  SN:chrX LN:156040895
## @PG  ID:hisat2   PN:hisat2   VN:2.2.0    CL:"/Users/dtang/github/rnaseq/hisat2/../src/hisat2-2.2.0/hisat2-align-s --wrapper basic-0 --dta -p 4 -x ../raw/chrX_data/indexes/chrX_tran -1 /tmp/4195.inpipe1 -2 /tmp/4195.inpipe2"
## @PG  ID:samtools PN:samtools PP:hisat2   VN:1.15 CL:samtools addreplacerg -r @RG\tID:ERR188273\tSM:ERR188273\tPL:illumina -o eg/ERR188273_chrX_rg.bam eg/ERR188273_chrX.bam
## @PG  ID:samtools.1   PN:samtools PP:samtools VN:1.15 CL:samtools addreplacerg -r @RG\tID:ERR188273_2\tSM:ERR188273_2\tPL:illumina_2 -o eg/ERR188273_chrX_rg2.bam eg/ERR188273_chrX_rg.bam
## @RG  ID:ERR188273_2  SM:ERR188273_2  PL:illumina_2
</code></pre>
<p>Popular alignment tools such as BWA MEM and STAR can add read groups;
use the <code>-R</code> and <code>--outSAMattrRGline</code> parameters for the respective
tool.</p>
<pre><code>bwa mem \
  -M \
  -t ${thread} \
  -R "@RG\tID:${sample_name}\tSM:${sample}\tPL:${platform}" \
  ${fasta} \
  ${fastq1} \
  ${fastq2} |
  samtools sort -@ ${thread} -O BAM |\
  tee ${sample_name}.bam |\
  samtools index - ${sample_name}.bam.bai

STAR \
  --runMode alignReads \
  --genomeDir ${star_index} \
  --readFilesIn ${fastq1} ${fastq2} \
  --readFilesCommand "gunzip -c" \
  --outFileNamePrefix ${prefix}. \
  --outSAMtype BAM Unsorted \
  --twopassMode Basic \
  --outSAMattrRGline ID:${id} PL:Illumina PU:${pu} LB:${lb} PI:0 SM:${sm} \
  --outSAMattributes NH HI AS nM NM ch \
  --runThreadN ${num_threads}
</code></pre>
<h2 id="interpreting-the-bam-flags">Interpreting the BAM flags</h2>
<p>The second column in a SAM/BAM file is the flag column; use the <code>flags</code>
subcommand to understand specific flags. They may seem confusing at
first but the encoding allows details about a read to be stored by just
using a few digits. The trick is to convert the numerical digit into
binary, and then use the table to interpret the binary numbers, where 1
= true and 0 = false. I wrote a blog post on BAM flags at
<a href="http://davetang.org/muse/2014/03/06/understanding-bam-flags/">http://davetang.org/muse/2014/03/06/understanding-bam-flags/</a>.</p>
<pre><code class="language-bash">samtools flags
</code></pre>
<pre><code>## About: Convert between textual and numeric flag representation
## Usage: samtools flags FLAGS...
## 
## Each FLAGS argument is either an INT (in decimal/hexadecimal/octal) representing
## a combination of the following numeric flag values, or a comma-separated string
## NAME,...,NAME representing a combination of the following flag names:
## 
##    0x1     1  PAIRED         paired-end / multiple-segment sequencing technology
##    0x2     2  PROPER_PAIR    each segment properly aligned according to aligner
##    0x4     4  UNMAP          segment unmapped
##    0x8     8  MUNMAP         next segment in the template unmapped
##   0x10    16  REVERSE        SEQ is reverse complemented
##   0x20    32  MREVERSE       SEQ of next segment in template is rev.complemented
##   0x40    64  READ1          the first segment in the template
##   0x80   128  READ2          the last segment in the template
##  0x100   256  SECONDARY      secondary alignment
##  0x200   512  QCFAIL         not passing quality controls or other filters
##  0x400  1024  DUP            PCR or optical duplicate
##  0x800  2048  SUPPLEMENTARY  supplementary alignment
</code></pre>
<p>Find out about a <code>73</code> flag.</p>
<pre><code class="language-bash">samtools flags 73
</code></pre>
<pre><code>## 0x49 73  PAIRED,MUNMAP,READ1
</code></pre>
<h2 id="filtering-unmapped-reads">Filtering unmapped reads</h2>
<p>Use <code>-F 4</code> to filter out unmapped reads.</p>
<pre><code class="language-bash">samtools view -F 4 -b eg/ERR188273_chrX.bam &gt; eg/ERR188273_chrX.mapped.bam
</code></pre>
<p>Use <code>-f 4</code> to keep only unmapped reads.</p>
<pre><code class="language-bash">samtools view -f 4 -b eg/ERR188273_chrX.bam &gt; eg/ERR188273_chrX.unmapped.bam
</code></pre>
<p>We can use the <code>flags</code> subcommand to confirm that a value of four
represents an unmapped read.</p>
<pre><code class="language-bash">samtools flags 4
</code></pre>
<pre><code>## 0x4  4   UNMAP
</code></pre>
<h3 id="proper-pair">Proper pair</h3>
<p>Reads that are properly paired are mapped within an expected distance
with each other and with one pair in the reverse complement orientation.
The script <code>generate_random_seq.pl</code> can generate reads that originate
from different references and are thus discordant and not properly
paired (as well as properly paired reads). In the example below, 10% of
reads are not properly paired (set with <code>-d 0.1</code>).</p>
<pre><code class="language-bash">script/generate_random_seq.pl 30 10000 1984 &gt; test_ref.fa
script/random_paired_end.pl -f test_ref.fa -l 100 -n 10000 -m 300 -d 0.1
bwa index test_ref.fa
bwa mem test_ref.fa l100_n10000_d300_1984_1.fq.gz l100_n10000_d300_1984_2.fq.gz &gt; aln.sam
</code></pre>
<pre><code>## [bwa_index] Pack FASTA... 0.00 sec
## [bwa_index] Construct BWT for the packed sequence...
## [bwa_index] 0.06 seconds elapse.
## [bwa_index] Update BWT... 0.00 sec
## [bwa_index] Pack forward-only FASTA... 0.00 sec
## [bwa_index] Construct SA from BWT and Occ... 0.03 sec
## [main] Version: 0.7.17-r1188
## [main] CMD: bwa index test_ref.fa
## [main] Real time: 0.096 sec; CPU: 0.090 sec
## [M::bwa_idx_load_from_disk] read 0 ALT contigs
## [M::process] read 20000 sequences (2000000 bp)...
## [M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (0, 9006, 0, 0)
## [M::mem_pestat] skip orientation FF as there are not enough pairs
## [M::mem_pestat] analyzing insert size distribution for orientation FR...
## [M::mem_pestat] (25, 50, 75) percentile: (399, 399, 399)
## [M::mem_pestat] low and high boundaries for computing mean and std.dev: (399, 399)
## [M::mem_pestat] mean and std.dev: (399.00, 0.00)
## [M::mem_pestat] low and high boundaries for proper pairs: (399, 399)
## [M::mem_pestat] skip orientation RF as there are not enough pairs
## [M::mem_pestat] skip orientation RR as there are not enough pairs
## [M::mem_process_seqs] Processed 20000 reads in 0.440 CPU sec, 0.440 real sec
## [main] Version: 0.7.17-r1188
## [main] CMD: bwa mem test_ref.fa l100_n10000_d300_1984_1.fq.gz l100_n10000_d300_1984_2.fq.gz
## [main] Real time: 0.486 sec; CPU: 0.480 sec
</code></pre>
<p><code>samtools flagstat</code> will indicate that some reads mapped to different
chromosomes.</p>
<pre><code class="language-bash">samtools flagstat aln.sam
</code></pre>
<pre><code>## 20000 + 0 in total (QC-passed reads + QC-failed reads)
## 20000 + 0 primary
## 0 + 0 secondary
## 0 + 0 supplementary
## 0 + 0 duplicates
## 0 + 0 primary duplicates
## 20000 + 0 mapped (100.00% : N/A)
## 20000 + 0 primary mapped (100.00% : N/A)
## 20000 + 0 paired in sequencing
## 10000 + 0 read1
## 10000 + 0 read2
## 18012 + 0 properly paired (90.06% : N/A)
## 20000 + 0 with itself and mate mapped
## 0 + 0 singletons (0.00% : N/A)
## 1988 + 0 with mate mapped to a different chr
## 1988 + 0 with mate mapped to a different chr (mapQ&gt;=5)
</code></pre>
<p>Flag of a proper pair.</p>
<pre><code class="language-bash">samtools flag $(samtools view -f 2 aln.sam | head -1 | cut -f2)
</code></pre>
<pre><code>## 0x63 99  PAIRED,PROPER_PAIR,MREVERSE,READ1
</code></pre>
<p>Flag of a pair (not is not a proper pair).</p>
<pre><code class="language-bash">samtools flag $(samtools view -F 2 aln.sam | head -1 | cut -f2)
</code></pre>
<pre><code>## 0x61 97  PAIRED,MREVERSE,READ1
</code></pre>
<h2 id="extracting-entries-mapping-to-a-specific-loci">Extracting entries mapping to a specific loci</h2>
<p>Use <code>samtools view</code> and the <code>ref:start-end</code> syntax to extract reads
mapping within a specific genomic loci; this requires a BAM index file.</p>
<pre><code class="language-bash">samtools view eg/ERR188273_chrX.bam chrX:20000-30000
</code></pre>
<pre><code>## ERR188273.4711308    73  chrX    21649   0   5S70M   =   21649   0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.4711308    133 chrX    21649   0   *   =   21649   0   CTACAGGTGCCCGCCACCATGCCCAGCTAATTTTTTTTGTATTTTTAGTAGAGATGGGGTTTCACTGTGTTGGCC CB@FDFFFHHGFHIJJJJIIIIIIIGGGIJGIIJJJJJJFFHIIIIGECHEHHGGHHFF?AACCDDDDDDDDBCD YT:Z:UP
</code></pre>
<p>Note that this takes into account the mapping of the entire read and not
just the starting position. For example, if you specified
chrX:20000-30000, a 75 bp long read that starts its mapping from
position 19999 will also be returned. In addition, you can save the
output as another BAM file if you want.</p>
<pre><code class="language-bash">samtools view -b eg/ERR188273_chrX.bam chrX:20000-30000 &gt; eg/ERR188273_chrX_20000_30000.bam
</code></pre>
<p>If you want reads mapped to a single reference (e.g. chromosome), just
specify the <code>ref</code> and leave out the start and end values.</p>
<pre><code class="language-bash">samtools view eg/ERR188273_chrX.bam chrX | head
</code></pre>
<pre><code>## ERR188273.4711308    73  chrX    21649   0   5S70M   =   21649   0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.4711308    133 chrX    21649   0   *   =   21649   0   CTACAGGTGCCCGCCACCATGCCCAGCTAATTTTTTTTGTATTTTTAGTAGAGATGGGGTTTCACTGTGTTGGCC CB@FDFFFHHGFHIJJJJIIIIIIIGGGIJGIIJJJJJJFFHIIIIGECHEHHGGHHFF?AACCDDDDDDDDBCD YT:Z:UP
## ERR188273.4711308    329 chrX    233717  0   5S70M   =   233717  0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.14904746   99  chrX    251271  60  75M =   251317  121 GAAAAATGGGCCCAGGGGACCGGCGCTCAGCATACAGAGGACCCGCGCCGGCACCTGCCTCTGAGTTCCCTTAGT @@&lt;DDDDDFB&gt;HHEGIIGAGIIIBGIIG@FECH&lt;F@GIIFAE=?BCBBCBBB5@&lt;?CBBCCCCAACDCCCCCCCC AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YS:i:-2 YT:Z:CP NH:i:1
## ERR188273.14904746   147 chrX    251317  60  75M =   251271  -121    GCCGGCCCCTGCCTCTGAGTTCCCTTAGTACTTATTGATCATTATCGGGGAGAGGGGGATGTGGCAGGACAATAG #######B?DAHC@EGIIGGEHHGC@GFBFCEGFCIGG@EG@@H&lt;JIEHEF@IGEHGHIIHFGHDDFDDDDD?&lt;B AS:i:-2 ZS:i:-7 XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:6A68   YS:i:0  YT:Z:CP NH:i:1
## ERR188273.5849805    163 chrX    265951  1   75M =   266022  146 CGGGTTCACGCCATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGACTACAGGCGCCCGCCACCACGCCCGGCTAA @CCFDFFFHHHHHJJJJJJFJJJJJIJIJJJJJJJJGHIJJJJJEHIJIJGIIJJJHHFFDDEDDDDDDDDDDDD AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YS:i:0  YT:Z:CP NH:i:2
## ERR188273.1232356    369 chrX    265984  1   75M =   118343251   0   GAGTAGCTGGGACTACAGGCGCCCGCCACCACGCCCGGCTAATTTTTTGTATTTTTAGTAGAGACGGGGTTTCAC @@CA@B&gt;DC&gt;&gt;+@::8-755-BBBFDDEHHBGGEGHEEIJIIGIJJIGEIIIJJJIIJJIGGHHHGGFFFFF@@C AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YT:Z:UP NH:i:10
## ERR188273.5927795    385 chrX    265991  1   75M =   114048277   0   TGGGACTACAGGCGCCCGCCACCACGCCCGGCTAATTTTTTGTATTTTTAGTAGAGACGGGGTTTCACCGTGTTA =?BB??BD?FBHHBEAE@CDGG@HH=FA@GEGE;FGACCHBE6?A=ACE9)7@DCE&gt;&gt;5'3=338:;:&gt;2&lt;AA?: AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YT:Z:UP NH:i:10
## ERR188273.5849805    83  chrX    266022  1   75M =   265951  -146    CTAATTTTTTGTATTTTTAGTAGAGACGGGGTTTCACCGTGTTAGCCAGGATGGTGTCGATCTCCTGACCTCGTG DDDDDDDEEEEEDBFEDGHHHHHHJHIJJIGIGHFBJJIHGJJIIJJJJJJJJIGIJJJJJJHHHHHDFFFFCCB AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YS:i:0  YT:Z:CP NH:i:2
## ERR188273.13655123   113 chrX    266022  1   75M =   118343234   0   CTAATTTTTTGTATTTTTAGTAGAGACGGGGTTTCACCGTGTTAGCCAGGATGGTGTCGATCTCCTGACCTCGTG AACBBBACCCC&gt;;3?BCFFEEHHHEEGIGGHAGFBBHFBHHEHCG@&lt;@ABG??@@?BB9GBGAFFD&lt;&lt;DDAD@@@ AS:i:0  ZS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YT:Z:UP NH:i:2
</code></pre>
<p>You can also use a BED file, with several entries, to extract reads of
interest.</p>
<pre><code class="language-bash">cat eg/my.bed 

samtools view -L eg/my.bed eg/ERR188273_chrX.bam
</code></pre>
<pre><code>## chrX 20000   30000
## chrX 233000  260000
## ERR188273.4711308    73  chrX    21649   0   5S70M   =   21649   0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.4711308    133 chrX    21649   0   *   =   21649   0   CTACAGGTGCCCGCCACCATGCCCAGCTAATTTTTTTTGTATTTTTAGTAGAGATGGGGTTTCACTGTGTTGGCC CB@FDFFFHHGFHIJJJJIIIIIIIGGGIJGIIJJJJJJFFHIIIIGECHEHHGGHHFF?AACCDDDDDDDDBCD YT:Z:UP
## ERR188273.4711308    329 chrX    233717  0   5S70M   =   233717  0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.14904746   99  chrX    251271  60  75M =   251317  121 GAAAAATGGGCCCAGGGGACCGGCGCTCAGCATACAGAGGACCCGCGCCGGCACCTGCCTCTGAGTTCCCTTAGT @@&lt;DDDDDFB&gt;HHEGIIGAGIIIBGIIG@FECH&lt;F@GIIFAE=?BCBBCBBB5@&lt;?CBBCCCCAACDCCCCCCCC AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YS:i:-2 YT:Z:CP NH:i:1
## ERR188273.14904746   147 chrX    251317  60  75M =   251271  -121    GCCGGCCCCTGCCTCTGAGTTCCCTTAGTACTTATTGATCATTATCGGGGAGAGGGGGATGTGGCAGGACAATAG #######B?DAHC@EGIIGGEHHGC@GFBFCEGFCIGG@EG@@H&lt;JIEHEF@IGEHGHIIHFGHDDFDDDDD?&lt;B AS:i:-2 ZS:i:-7 XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:6A68   YS:i:0  YT:Z:CP NH:i:1
</code></pre>
<h2 id="extracting-only-the-first-read-from-paired-end-bam-files">Extracting only the first read from paired end BAM files</h2>
<p>Sometimes you only want the first pair of a mate. 0x0040 is hexadecimal
for 64 (i.e. 16 * 4), which is binary for 1000000, corresponding to the
read in the first read pair.</p>
<pre><code class="language-bash">samtools view -b -f 0x0040 eg/ERR188273_chrX.bam &gt; eg/first.bam
</code></pre>
<p>Once again, you can use <code>flags</code> to verify this (it also accepts
hexadecimal input).</p>
<pre><code class="language-bash">samtools flags 0x0040
</code></pre>
<pre><code>## 0x40 64  READ1
</code></pre>
<h2 id="stats">Stats</h2>
<p>For simple statistics use <code>samtools flagstat</code>.</p>
<pre><code class="language-bash">samtools flagstat eg/ERR188273_chrX.bam
</code></pre>
<pre><code>## 1176360 + 0 in total (QC-passed reads + QC-failed reads)
## 1160084 + 0 primary
## 16276 + 0 secondary
## 0 + 0 supplementary
## 0 + 0 duplicates
## 0 + 0 primary duplicates
## 1126961 + 0 mapped (95.80% : N/A)
## 1110685 + 0 primary mapped (95.74% : N/A)
## 1160084 + 0 paired in sequencing
## 580042 + 0 read1
## 580042 + 0 read2
## 1060858 + 0 properly paired (91.45% : N/A)
## 1065618 + 0 with itself and mate mapped
## 45067 + 0 singletons (3.88% : N/A)
## 0 + 0 with mate mapped to a different chr
## 0 + 0 with mate mapped to a different chr (mapQ&gt;=5)
</code></pre>
<p>For more stats, use <code>samtools stats</code>.</p>
<pre><code class="language-bash">samtools stats eg/ERR188273_chrX.bam | grep ^SN
</code></pre>
<pre><code>## SN   raw total sequences:    1160084 # excluding supplementary and secondary reads
## SN   filtered sequences: 0
## SN   sequences:  1160084
## SN   is sorted:  1
## SN   1st fragments:  580042
## SN   last fragments: 580042
## SN   reads mapped:   1110685
## SN   reads mapped and paired:    1065618 # paired-end technology bit set + both mates mapped
## SN   reads unmapped: 49399
## SN   reads properly paired:  1060858 # proper-pair bit set
## SN   reads paired:   1160084 # paired-end technology bit set
## SN   reads duplicated:   0   # PCR or optical duplicate bit set
## SN   reads MQ0:  905 # mapped and MQ=0
## SN   reads QC failed:    0
## SN   non-primary alignments: 16276
## SN   supplementary alignments:   0
## SN   total length:   87006300    # ignores clipping
## SN   total first fragment length:    43503150    # ignores clipping
## SN   total last fragment length: 43503150    # ignores clipping
## SN   bases mapped:   83301375    # ignores clipping
## SN   bases mapped (cigar):   83064942    # more accurate
## SN   bases trimmed:  0
## SN   bases duplicated:   0
## SN   mismatches: 423271  # from NM fields
## SN   error rate: 5.095663e-03    # mismatches / bases mapped (cigar)
## SN   average length: 75
## SN   average first fragment length:  75
## SN   average last fragment length:   75
## SN   maximum length: 75
## SN   maximum first fragment length:  75
## SN   maximum last fragment length:   75
## SN   average quality:    36.0
## SN   insert size average:    182.7
## SN   insert size standard deviation: 176.0
## SN   inward oriented pairs:  530763
## SN   outward oriented pairs: 1042
## SN   pairs with other orientation:   1004
## SN   pairs on different chromosomes: 0
## SN   percentage of properly paired reads (%):    91.4
</code></pre>
<h2 id="samtools-calmdfillmd">samtools calmd/fillmd</h2>
<p>The <code>calmd</code> or <code>fillmd</code> tool is useful for visualising mismatches and
insertions in an alignment of a read to a reference genome. The <code>-e</code>
argument changes identical bases between the read and reference into
<code>=</code>.</p>
<pre><code class="language-bash">samtools view -b eg/ERR188273_chrX.bam | samtools fillmd -e - genome/chrX.fa &gt; eg/ERR188273_chrX_fillmd.bam

head eg/ERR188273_chrX_fillmd.bam
</code></pre>
<pre><code>## @HD  VN:1.0  SO:coordinate
## @SQ  SN:chrX LN:156040895
## @PG  ID:hisat2   PN:hisat2   VN:2.2.0    CL:"/Users/dtang/github/rnaseq/hisat2/../src/hisat2-2.2.0/hisat2-align-s --wrapper basic-0 --dta -p 4 -x ../raw/chrX_data/indexes/chrX_tran -1 /tmp/4195.inpipe1 -2 /tmp/4195.inpipe2"
## @PG  ID:samtools PN:samtools PP:hisat2   VN:1.15 CL:samtools view -b eg/ERR188273_chrX.bam
## @PG  ID:samtools.1   PN:samtools PP:samtools VN:1.15 CL:samtools fillmd -e - genome/chrX.fa
## ERR188273.4711308    73  chrX    21649   0   5S70M   =   21649   0   CGGGT====================================================================== @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.4711308    133 chrX    21649   0   *   =   21649   0   CTACAGGTGCCCGCCACCATGCCCAGCTAATTTTTTTTGTATTTTTAGTAGAGATGGGGTTTCACTGTGTTGGCC CB@FDFFFHHGFHIJJJJIIIIIIIGGGIJGIIJJJJJJFFHIIIIGECHEHHGGHHFF?AACCDDDDDDDDBCD YT:Z:UP
## ERR188273.4711308    329 chrX    233717  0   5S70M   =   233717  0   CGGGT====================================================================== @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.14904746   99  chrX    251271  60  75M =   251317  121 =========================================================================== @@&lt;DDDDDFB&gt;HHEGIIGAGIIIBGIIG@FECH&lt;F@GIIFAE=?BCBBCBBB5@&lt;?CBBCCCCAACDCCCCCCCC AS:i:0  XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:75 YS:i:-2 YT:Z:CP NH:i:1
## ERR188273.14904746   147 chrX    251317  60  75M =   251271  -121    ======C==================================================================== #######B?DAHC@EGIIGGEHHGC@GFBFCEGFCIGG@EG@@H&lt;JIEHEF@IGEHGHIIHFGHDDFDDDDD?&lt;B AS:i:-2 ZS:i:-7 XN:i:0  XM:i:1  XO:i:0  XG:i:0  NM:i:1  MD:Z:6A68   YS:i:0  YT:Z:CP NH:i:1
</code></pre>
<h2 id="creating-fastq-files-from-a-bam-file">Creating FASTQ files from a BAM file</h2>
<p>Use the <code>fastq</code> tool to create FASTQ files from a BAM file. For
paired-end reads, use <code>-1</code> and <code>-2</code> to create separate FASTA files.</p>
<pre><code class="language-bash">samtools fastq -1 eg/ERR188273_chrX_1.fq -2 eg/ERR188273_chrX_2.fq eg/ERR188273_chrX.bam
head eg/ERR188273_chrX_1.fq
</code></pre>
<pre><code>## [M::bam2fq_mainloop] discarded 0 singletons
## [M::bam2fq_mainloop] processed 1160084 reads
## @ERR188273.4711308
## CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA
## +
## @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@
## @ERR188273.14904746
## GAAAAATGGGCCCAGGGGACCGGCGCTCAGCATACAGAGGACCCGCGCCGGCACCTGCCTCTGAGTTCCCTTAGT
## +
## @@&lt;DDDDDFB&gt;HHEGIIGAGIIIBGIIG@FECH&lt;F@GIIFAE=?BCBBCBBB5@&lt;?CBBCCCCAACDCCCCCCCC
## @ERR188273.5849805
## CACGAGGTCAGGAGATCGACACCATCCTGGCTAACACGGTGAAACCCCGTCTCTACTAAAAATACAAAAAATTAG
</code></pre>
<h2 id="random-subsampling-of-bam-file">Random subsampling of BAM file</h2>
<p>The SAMtools view <code>-s</code> parameter allows you to randomly sub-sample a BAM
file. Using <code>-s 0.5</code> will create a new BAM file with a random half of
all mapped reads; unmapped reads are not sampled.</p>
<pre><code class="language-bash">samtools view -s 0.5 -b eg/ERR188273_chrX.bam &gt; eg/ERR188273_chrX_rand.bam
</code></pre>
<h2 id="count-number-of-reads">Count number of reads</h2>
<p>Use <code>samtools idxstats</code> to print stats on a BAM file; this requires an
index file which is created by running <code>samtools index</code>. The output of
idxstats is a file with four tab-delimited columns:</p>
<ol>
<li>Reference name</li>
<li>Sequence length of reference</li>
<li>Number of mapped reads</li>
<li>Number of unmapped reads</li>
</ol>
<!-- end list -->

<pre><code class="language-bash">samtools idxstats eg/ERR188273_chrX.bam
</code></pre>
<pre><code>## chrX 156040895   1126961 45067
## *    0   0   4332
</code></pre>
<p>We can use this with <code>awk</code> to calculate:</p>
<p>The number of mapped reads by summing the third column.</p>
<pre><code class="language-bash">samtools idxstats eg/ERR188273_chrX.bam  | awk '{s+=$3} END {print s}'
</code></pre>
<pre><code>## 1126961
</code></pre>
<p>The number of reads, which is the sum of mapped and unmapped reads.</p>
<pre><code class="language-bash">samtools idxstats eg/ERR188273_chrX.bam | awk '{s+=$3+$4} END {print s}'
</code></pre>
<pre><code>## 1176360
</code></pre>
<h2 id="obtaining-genomic-sequence">Obtaining genomic sequence</h2>
<p>Use <code>faidx</code> to fetch genomic sequence; coordinates are 1-based.</p>
<p>We need to first index the reference FASTA file that was used to map the
reads.</p>
<pre><code class="language-bash">samtools faidx genome/chrX.fa
</code></pre>
<p>Now we can obtain the sequence.</p>
<pre><code class="language-bash">samtools faidx genome/chrX.fa chrX:300000-300100
</code></pre>
<pre><code>## &gt;chrX:300000-300100
## ctgagatcgtgccactgcactccagcctgggcgacagagcgagactccatctcaaaaaaa
## aaaaaaaaaaaaaagaTggggtctctctatgttggccaggt
</code></pre>
<h2 id="comparing-bam-files">Comparing BAM files</h2>
<p>Install <a href="https://deeptools.readthedocs.io/en/develop/">deepTools</a> and
use
<a href="https://deeptools.readthedocs.io/en/develop/content/tools/bamCompare.html">bamCompare</a>.
The bigWig output file shows the ratio of reads between <code>b1</code> and <code>b2</code> in
50 bp (default) windows.</p>
<h2 id="converting-reference-names">Converting reference names</h2>
<p>One of the most annoying bioinformatics problems is the use of different
chromosome names, e.g. chr1 vs 1, in different references even when the
sequences are identical. The GRCh38 reference downloaded from Ensembl
has chromosome names without the <code>chr</code>:</p>
<pre><code>&gt;1 dna:chromosome chromosome:GRCh38:1:1:248956422:1 REF
</code></pre>
<p>Whereas the reference names from UCSC has the <code>chr</code>:</p>
<pre><code>&gt;chr1  AC:CM000663.2  gi:568336023  LN:248956422  rl:Chromosome  M5:6aef897c3d6ff0c78aff06ac189178dd  AS:GRCh38
</code></pre>
<p>Luckily you can change the reference names using <code>samtools reheader</code> but
just make sure your reference sequences are actually identical.</p>
<pre><code class="language-bash">samtools view eg/ERR188273_chrX.bam | head -2
</code></pre>
<pre><code>## ERR188273.4711308    73  chrX    21649   0   5S70M   =   21649   0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.4711308    133 chrX    21649   0   *   =   21649   0   CTACAGGTGCCCGCCACCATGCCCAGCTAATTTTTTTTGTATTTTTAGTAGAGATGGGGTTTCACTGTGTTGGCC CB@FDFFFHHGFHIJJJJIIIIIIIGGGIJGIIJJJJJJFFHIIIIGECHEHHGGHHFF?AACCDDDDDDDDBCD YT:Z:UP
</code></pre>
<p>View header</p>
<pre><code class="language-bash">samtools view -H eg/ERR188273_chrX.bam
</code></pre>
<pre><code>## @HD  VN:1.0  SO:coordinate
## @SQ  SN:chrX LN:156040895
## @PG  ID:hisat2   PN:hisat2   VN:2.2.0    CL:"/Users/dtang/github/rnaseq/hisat2/../src/hisat2-2.2.0/hisat2-align-s --wrapper basic-0 --dta -p 4 -x ../raw/chrX_data/indexes/chrX_tran -1 /tmp/4195.inpipe1 -2 /tmp/4195.inpipe2"
## @PG  ID:samtools PN:samtools PP:hisat2   VN:1.15 CL:samtools view -H eg/ERR188273_chrX.bam
</code></pre>
<p>Substitute header with new name.</p>
<pre><code class="language-bash">samtools view -H eg/ERR188273_chrX.bam | sed 's/SN:chrX/SN:X/' &gt; eg/my_header
</code></pre>
<p>Save bam file with new ref and check it out.</p>
<pre><code class="language-bash">samtools reheader eg/my_header eg/ERR188273_chrX.bam &gt; eg/ERR188273_X.bam
samtools view eg/ERR188273_X.bam | head -2
</code></pre>
<pre><code>## ERR188273.4711308    73  X   21649   0   5S70M   =   21649   0   CGGGTGATCACGAGGTCAGGAGATCAAGACCATCCTGGCCAACACAGTGAAACCCCATCTCTACTAAAAATACAA @@@F=DDFFHGHBHIFFHIGGIFGEGHFHIGIGIFIIIGIGIGGDHIIGIIC@&gt;DGHCHHHGHHFFFFFDEACC@ AS:i:-5 ZS:i:-5 XN:i:0  XM:i:0  XO:i:0  XG:i:0  NM:i:0  MD:Z:70 YT:Z:UP NH:i:2
## ERR188273.4711308    133 X   21649   0   *   =   21649   0   CTACAGGTGCCCGCCACCATGCCCAGCTAATTTTTTTTGTATTTTTAGTAGAGATGGGGTTTCACTGTGTTGGCC CB@FDFFFHHGFHIJJJJIIIIIIIGGGIJGIIJJJJJJFFHIIIIGECHEHHGGHHFF?AACCDDDDDDDDBCD YT:Z:UP
</code></pre>
<h2 id="coverage">Coverage</h2>
<p>There are several ways to calculate coverage, i.e. count the number of
bases mapped to positions on the reference. <code>samtools depth</code> will return
three columns: reference, position, and coverage.</p>
<pre><code class="language-bash">samtools depth eg/ERR188273_chrX.bam | head
</code></pre>
<pre><code>## chrX 21649   1
## chrX 21650   1
## chrX 21651   1
## chrX 21652   1
## chrX 21653   1
## chrX 21654   1
## chrX 21655   1
## chrX 21656   1
## chrX 21657   1
## chrX 21658   1
</code></pre>
<p>The <code>samtools mpileup</code> command can also provide depth information (but
not for reads that have a mapping quality of 0, by default) with some
additional information:</p>
<ol>
<li>Sequence name</li>
<li>1-based coordinate</li>
<li>Reference base (when used with <code>-f</code>)</li>
<li>Number of reads covering this position</li>
<li>Read bases</li>
<li>Base qualities</li>
<li>Alignment mapping qualities (when used with <code>-s</code>)</li>
</ol>
<!-- end list -->

<pre><code class="language-bash">samtools mpileup -f genome/chrX.fa -s eg/ERR188273_chrX.bam | head
</code></pre>
<pre><code>## [mpileup] 1 samples in 1 input files
## chrX 251271  g   1   ^]. @   ]
## chrX 251272  a   1   .   @   ]
## chrX 251273  a   1   .   &lt;   ]
## chrX 251274  a   1   .   D   ]
## chrX 251275  a   1   .   D   ]
## chrX 251276  a   1   .   D   ]
## chrX 251277  t   1   .   D   ]
## chrX 251278  g   1   .   D   ]
## chrX 251279  g   1   .   F   ]
## chrX 251280  g   1   .   B   ]
</code></pre>
<p>Note that the start of the <code>samtools mpileup</code> output differ from the
start of the <code>samtools depth</code> output. This is because <code>mpileup</code> performs
some filtering by default. In the case of this example, read pairs that
are not both mapped will be ignored. To count these “orphan” reads, use
the <code>--count-orphans</code> argument.</p>
<pre><code class="language-bash">samtools mpileup -f genome/chrX.fa --count-orphans -s eg/ERR188273_chrX.bam | head
</code></pre>
<pre><code>## [mpileup] 1 samples in 1 input files
## chrX 21649   g   0   *   *   *
## chrX 21650   a   1   .   D   !
## chrX 21651   t   1   .   F   !
## chrX 21652   c   1   .   F   !
## chrX 21653   a   1   .   H   !
## chrX 21654   c   1   .   G   !
## chrX 21655   g   1   .   H   !
## chrX 21656   a   1   .   B   !
## chrX 21657   g   1   .   H   !
## chrX 21658   g   1   .   I   !
</code></pre>
<p>In addition <code>mpileup</code> performs “per-Base Alignment Quality” (BAQ) by
default and will adjust base quality scores. The default behaviour to to
skip bases with baseQ/BAQ smaller than 13. If you are finding
discrepancies between <code>mpileup</code>’s coverage calculation with another
coverage tool, you can either set <code>--min-BQ</code> to <code>0</code> or use <code>--no-BAQ</code> to
disable BAQ.</p>
<p>I have an <a href="https://davetang.org/muse/2015/08/26/samtools-mpileup/">old blog
post</a> on using
<code>mpileup</code>.</p>
<p><code>samtools coverage</code> will provide the following coverage statistics:</p>
<ol>
<li>rname - Reference name / chromosome</li>
<li>startpos - Start position</li>
<li>endpos - End position (or sequence length)</li>
<li>numreads - Number reads aligned to the region (after filtering)</li>
<li>covbases - Number of covered bases with depth >= 1</li>
<li>coverage - Proportion of covered bases [0..1]</li>
<li>meandepth - Mean depth of coverage</li>
<li>meanbaseq - Mean base quality in covered region</li>
<li>meanmapq - Mean mapping quality of selected reads</li>
</ol>
<!-- end list -->

<pre><code class="language-bash">samtools coverage eg/ERR188273_chrX.bam
</code></pre>
<pre><code>## #rname   startpos    endpos  numreads    covbases    coverage    meandepth   meanbaseq   meanmapq
## chrX 1   156040895   1110685 3402037 2.18022 0.532299    36.3    59.4
</code></pre>
<h2 id="stargazers-over-time">Stargazers over time</h2>
<p><a href="https://starchart.cc/davetang/learning_bam_file"><img alt="Stargazers over
time" src="https://starchart.cc/davetang/learning_bam_file.svg" /></a></p>
              
            </div>
          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
    
  </span>
</div>
    <script>var base_url = '.';</script>
    <script src="js/theme_extra.js" defer></script>
    <script src="js/theme.js" defer></script>
      <script src="search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>

<!--
MkDocs version : 1.2.3
Build Date UTC : 2022-04-18 08:37:31.475396+00:00
-->
